### PR TITLE
(Chapter 18) A few changes

### PR DESCRIPTION
**Pull request by ibeauregard**

### CODE REVIEW

File: 18_reinforcement_learning.ipynb [hunk 1]
```diff
@@ -1849,8 +1849,7 @@
     "    keras.layers.Dense(n_outputs)\n",
     "])\n",
     "\n",
-    "target = keras.models.clone_model(model)\n",
-    "target.set_weights(model.get_weights())"
+    "target = keras.models.clone_model(model)"
```
Comment (ageron) [inline 18_reinforcement_learning.ipynb:1852]: Surprisingly `clone_model()` does **not** copy the weights, it just copies the model architecture. Therefore the `set_weights()` call is actually required.

Here's some proof. Try running this code:

```python
import numpy as np
import tensorflow as tf
from tensorflow import keras

model = keras.models.Sequential([
    keras.layers.Dense(10, activation="relu"),
    keras.layers.Dense(1)
])
model.compile(loss="mse", optimizer="adam")
X_train = tf.random.uniform((100, 5))
y_train = tf.random.uniform((100, 1))
history = model.fit(X_train, y_train)

target = keras.models.clone_model(model)

for w1, w2 in zip(target.weights, model.weights):
  print(np.allclose(w1, w2))
```

File: 18_reinforcement_learning.ipynb [hunk 2]
```diff
@@ -1925,12 +1924,12 @@
     "        best_weights = model.get_weights()\n",
     "        best_score = step\n",
     "    print(\"\\rEpisode: {}, Steps: {}, eps: {:.3f}\".format(episode, step + 1, epsilon), end=\"\")\n",
-    "    if episode > 50:\n",
-    "        training_step(batch_size)\n",
     "    if episode % 50 == 0:\n",
     "        target.set_weights(model.get_weights())\n",
+    "    if episode >= 50:\n",
```
Comment (ageron) [inline 18_reinforcement_learning.ipynb:1929]: I'm not sure why `set_weights()` needs to happen before the `training_set()`? Could you please explain this?

**Review by ageron — CHANGES_REQUESTED** [review]

File: 18_reinforcement_learning.ipynb (hunk 1 above)
Comment (ibeauregard) [reply 18_reinforcement_learning.ipynb:1852]: The reason I deleted this line is because of lines 1930 and 1931: `if episode % 50 == 0: target.set_weights(model.get_weights())`. I thought that because of that, we could do without line 1853.

File: 18_reinforcement_learning.ipynb [hunk 3]
```diff
@@ -2025,8 +2024,7 @@
     "Q_values = state_values + advantages\n",
     "model = keras.models.Model(inputs=[input_states], outputs=[Q_values])\n",
     "\n",
-    "target = keras.models.clone_model(model)\n",
-    "target.set_weights(model.get_weights())"
+    "target = keras.models.clone_model(model)"
```
Comment (ibeauregard) [inline 18_reinforcement_learning.ipynb:2027]: Same justification as line 1853.

File: 18_reinforcement_learning.ipynb [hunk 4]
```diff
@@ -1757,7 +1757,7 @@
     "        best_weights = model.get_weights() # Not shown\n",
     "        best_score = step # Not shown\n",
     "    print(\"\\rEpisode: {}, Steps: {}, eps: {:.3f}\".format(episode, step + 1, epsilon), end=\"\") # Not shown\n",
-    "    if episode > 50:\n",
+    "    if episode >= 50:\n",
```
Comment (ibeauregard) [inline 18_reinforcement_learning.ipynb:1760]: I know you didn't ask for clarifications here, but allow me to go ahead. The book says we start training after 50 episodes. That means we have to start at `episode = 50` (0-based), not `episode = 51` like previously.

File: 18_reinforcement_learning.ipynb (hunk 2 above)
Comment (ibeauregard) [reply 18_reinforcement_learning.ipynb:1929]: Suppose`episode = 50`. Then (as per my comment on line 1760) we will call `training_step(batch_size)` for the first time. But first, we have to call `target.set_weights(model.get_weights())`.

File: 18_reinforcement_learning.ipynb (hunk 4 above)
Comment (ageron) [reply 18_reinforcement_learning.ipynb:1760]: Yes, I guessed as much, good catch! 😊

File: 18_reinforcement_learning.ipynb (hunk 2 above)
Comment (ageron) [reply 18_reinforcement_learning.ipynb:1929]: Ooooh, I get it now, thanks for the explanation. 🙏  Yes, this would work fine.

Okay, so if the goal is to avoid the redundant call to `set_weights()` at the very beginning of training, I think I'd prefer a smaller change, replacing:

```python
    if episode >= 50:
        training_step(batch_size)
    if episode % 50 == 0:
        target.set_weights(model.get_weights())
```

with:

```python
    if episode >= 50:
        training_step(batch_size)
        if episode % 50 == 0:
            target.set_weights(model.get_weights())
```

I've just indented the last 2 lines. In Python can you fix a program by just adding a few spaces! 😆 

Other than changing slightly less code, I think it's safer to keep `clone_model()` and `set_weights()` next to each other when the intention is to clone both a model's architecture and its weights. Or else we have model with the right architecture but invalid weights, and it's easy to use the invalid model by mistake, before it is correctly initialized.

Of course your solution avoids this issue, but there's still a risk. For example, if someone wants to try soft updates, they will presumably comment out the current code and uncomment the soft updates section like this:

```python
    #if episode % 50 == 0:
    #    target.set_weights(model.get_weights())
    #if episode >= 50:
    #    training_step(batch_size)
    # Alternatively, you can do soft updates at each step:
    if episode >= 50:
        target_weights = target.get_weights()
        online_weights = model.get_weights()
        for index in range(len(target_weights)):
            target_weights[index] = 0.99 * target_weights[index] + 0.01 * online_weights[index]
        target.set_weights(target_weights)
```

Note that we're now using `target.get_weights()` before it's ever properly initialized.

If you agree, then could you please update the PR to keep just the `>=50` changes? If it's too much hassle, don't worry, I can do it.

Comment (ageron) [reply 18_reinforcement_learning.ipynb:1929]: Sorry, I meant keep just the `>=50` changes & the indentation changes.

File: 18_reinforcement_learning.ipynb [hunk 5]
```diff
@@ -1925,12 +1925,12 @@
     "        best_weights = model.get_weights()\n",
     "        best_score = step\n",
     "    print(\"\\rEpisode: {}, Steps: {}, eps: {:.3f}\".format(episode, step + 1, epsilon), end=\"\")\n",
-    "    if episode > 50:\n",
+    "    if episode >= 50:\n",
+    "        if episode % 50 == 0:\n",
+    "            target.set_weights(model.get_weights())\n",
```
Comment (ibeauregard) [inline 18_reinforcement_learning.ipynb:1930]: Prior to my `<=` changes, `target.set_weights` was called at `episode == 50` and `training_step(batch_size)` was called at `episode == 51` (i.e. after). I thought we might want to keep that order.

File: 18_reinforcement_learning.ipynb [hunk 6]
```diff
@@ -2101,10 +2101,10 @@
     "        best_weights = model.get_weights()\n",
     "        best_score = step\n",
     "    print(\"\\rEpisode: {}, Steps: {}, eps: {:.3f}\".format(episode, step + 1, epsilon), end=\"\")\n",
-    "    if episode > 50:\n",
+    "    if episode >= 50:\n",
```
Comment (ibeauregard) [inline 18_reinforcement_learning.ipynb:2104]: Just to be consistent with the code around line 1930.

**Review by ibeauregard — COMMENTED** [review]
@ageron, I've applied the changes that you requested, except for a small detail. See my comments at lines 1930 and 2104.

**Comment by ageron** [comment]
Hi @ibeauregard ,
I've made a few improvements in the notebook, including tuning the learning rate for the DQN and its variants (it really improves training significantly). I also improved the Breakout example by automatically pressing FIRE (action 1) when the game starts and after each life lost. This also really improves training. While I was at it, I replaced `if episode > 50` with `if episode >= 50`, and I indented the `if episode % 50` blocks, so I'm closing this PR.
Once again, thanks a **lot** for your help! 👍
